File: algorithms/0020.valid-parentheses/solution.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  stack = []
  brackets = { '(' => ')', '{' => '}', '[' => ']' }

  s.each_char do |c|
    if brackets.has_key?(c)
      stack.push(c)
    elsif brackets.has_value?(c)
      return false if brackets[stack.pop] != c
    else
      return false
    end
  end

  stack.empty?
end

File: algorithms/0020.valid-parentheses/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec![];
        let brackets: HashMap<char, char> = HashMap::from([('(', ')'), ('{', '}'), ('[', ']')]);
        let close_brackets: Vec<char> = vec![')', '}', ']'];

        for c in s.chars() {
            if brackets.contains_key(&c) {
                stack.push(c);
            } else if close_brackets.contains(&c) {
                match stack.pop() {
                    Some(x) => {
                        if brackets[&x] != c {
                            return false;
                        }
                    }
                    _ => return false,
                }
            } else {
                return false;
            }
        }

        stack.is_empty()
    }
}
